File: backend/internal/model/table.go
package model

import "time"

type Table struct {
	BaseModel
	StoreID      uint      `gorm:"not null;index:idx_store_table" json:"store_id"`
	TableNo      string    `gorm:"size:20;not null;index:idx_store_table" json:"table_no"`
	Name         string    `gorm:"size:50" json:"name"`
	Type         string    `gorm:"size:20;default:normal" json:"type"`
	Capacity     int       `gorm:"default:4" json:"capacity"`
	Floor        int       `gorm:"default:1" json:"floor"`
	Area         string    `gorm:"size:50" json:"area"`
	QRCode       string    `gorm:"size:500" json:"qr_code"`
	QRCodeUrl    string    `gorm:"size:500" json:"qr_code_url"`
	Status       int       `gorm:"default:1" json:"status"`
	CurrentOrderID uint    `gorm:"index" json:"current_order_id"`
	CurrentCustomerCount int `gorm:"default:0" json:"current_customer_count"`
	CheckinTime  *time.Time `json:"checkin_time"`
	OccupiedDuration int  `gorm:"default:0" json:"occupied_duration"`
	TotalOrders  int       `gorm:"default:0" json:"total_orders"`
	TotalAmount  float64   `gorm:"default:0" json:"total_amount"`
	Store        Store     `gorm:"foreignKey:StoreID" json:"store,omitempty"`
}

type TableReservation struct {
	BaseModel
	StoreID      uint       `gorm:"not null;index" json:"store_id"`
	TableID      uint       `gorm:"index" json:"table_id"`
	MemberID     uint       `gorm:"index" json:"member_id"`
	MemberName   string     `gorm:"size:50" json:"member_name"`
	MemberPhone  string     `gorm:"size:20" json:"member_phone"`
	TableNo      string     `gorm:"size:20" json:"table_no"`
	ReserveDate  string     `gorm:"size:20;not null" json:"reserve_date"`
	ReserveTime  string     `gorm:"size:20;not null" json:"reserve_time"`
	PeopleCount  int        `gorm:"default:2" json:"people_count"`
	Status       int        `gorm:"default:1" json:"status"`
	CheckinStatus int       `gorm:"default:0" json:"checkin_status"`
	CheckinTime  *time.Time `json:"checkin_time"`
	CancelTime   *time.Time `json:"cancel_time"`
	Remark       string     `gorm:"size:500" json:"remark"`
	Source       string     `gorm:"size:20;default:wechat" json:"source"`
	OrderID      uint       `json:"order_id"`
	Table        Table      `gorm:"foreignKey:TableID" json:"table,omitempty"`
	Member       Member     `gorm:"foreignKey:MemberID" json:"member,omitempty"`
}

type QueueNumber struct {
	BaseModel
	StoreID      uint       `gorm:"not null;index" json:"store_id"`
	QueueType    string     `gorm:"size:20;default:small" json:"queue_type"`
	QueueNumber  string     `gorm:"size:20;not null" json:"queue_number"`
	Sequence     int        `gorm:"not null" json:"sequence"`
	MemberID     uint       `gorm:"index" json:"member_id"`
	MemberName   string     `gorm:"size:50" json:"member_name"`
	MemberPhone  string     `gorm:"size:20" json:"member_phone"`
	PeopleCount  int        `gorm:"default:2" json:"people_count"`
	Status       int        `gorm:"default:1" json:"status"`
	CallCount    int        `gorm:"default:0" json:"call_count"`
	LastCallTime *time.Time `json:"last_call_time"`
	CallTime     *time.Time `json:"call_time"`
	ArriveTime   *time.Time `json:"arrive_time"`
	CancelTime   *time.Time `json:"cancel_time"`
	AheadCount   int        `gorm:"default:0" json:"ahead_count"`
	WaitDuration int        `gorm:"default:0" json:"wait_duration"`
	Remark       string     `gorm:"size:200" json:"remark"`
	TableID      uint       `json:"table_id"`
	TableNo      string     `gorm:"size:20" json:"table_no"`
	Member       Member     `gorm:"foreignKey:MemberID" json:"member,omitempty"`
}

type QueueConfig struct {
	BaseModel
	StoreID       uint   `gorm:"not null;uniqueIndex" json:"store_id"`
	SmallPrefix   string `gorm:"size:10;default:A" json:"small_prefix"`
	SmallCapacity int    `gorm:"default:4" json:"small_capacity"`
	MediumPrefix  string `gorm:"size:10;default:B" json:"medium_prefix"`
	MediumCapacity int   `gorm:"default:6" json:"medium_capacity"`
	LargePrefix   string `gorm:"size:10;default:C" json:"large_prefix"`
	LargeCapacity  int   `gorm:"default:10" json:"large_capacity"`
	AutoCall      bool   `gorm:"default:true" json:"auto_call"`
	CallInterval  int    `gorm:"default:300" json:"call_interval"`
	MaxCallCount  int    `gorm:"default:3" json:"max_call_count"`
	AutoExpire    bool   `gorm:"default:true" json:"auto_expire"`
	ExpireMinutes int    `gorm:"default:15" json:"expire_minutes"`
	VoiceNotify   bool   `gorm:"default:true" json:"voice_notify"`
	SMSNotify     bool   `gorm:"default:false" json:"sms_notify"`
}

type TableArea struct {
	BaseModel
	StoreID      uint   `gorm:"not null;index" json:"store_id"`
	Name         string `gorm:"size:50;not null" json:"name"`
	SortOrder    int    `gorm:"default:0" json:"sort_order"`
	Status       int    `gorm:"default:1" json:"status"`
}

File: miniprogram/src/services/table.ts
import request from './request'

export interface TableInfo {
  id: number
  storeId: number
  storeName: string
  tableNo: string
  tableType: string
  capacity: number
  area: string
  floor: number
  status: number
  message: string
}

export interface TableItem {
  id: number
  storeId: number
  tableNo: string
  name: string
  type: string
  capacity: number
  floor: number
  area: string
  qrCode: string
  qrCodeUrl: string
  status: number
  currentOrderId: number
  currentCustomerCount: number
  checkinTime: string
}

export interface ReservationItem {
  id: number
  storeId: number
  tableId: number
  memberId: number
  memberName: string
  memberPhone: string
  tableNo: string
  reserveDate: string
  reserveTime: string
  peopleCount: number
  status: number
  checkinStatus: number
  checkinTime: string
  cancelTime: string
  remark: string
  source: string
  orderId: number
  createdAt: string
}

export interface TimeSlot {
  time: string
  available: number
  total: number
  status: number
}

export interface QueueItem {
  id: number
  storeId: number
  queueType: string
  queueNumber: string
  sequence: number
  memberId: number
  memberName: string
  memberPhone: string
  peopleCount: number
  status: number
  callCount: number
  lastCallTime: string
  callTime: string
  arriveTime: string
  cancelTime: string
  aheadCount: number
  waitDuration: number
  remark: string
  tableId: number
  tableNo: string
  createdAt: string
}

export interface QueueStatus {
  queueNumber: string
  queueType: string
  status: number
  sequence: number
  aheadCount: number
  waitTime: number
  peopleCount: number
  createdAt: string
}

export interface QueueConfig {
  storeId: number
  smallPrefix: string
  smallCapacity: number
  mediumPrefix: string
  mediumCapacity: number
  largePrefix: string
  largeCapacity: number
  autoCall: boolean
  callInterval: number
  maxCallCount: number
  autoExpire: boolean
  expireMinutes: number
  voiceNotify: boolean
  smsNotify: boolean
}

export const tableApi = {
  scanQRCode: (scene: string): Promise<TableInfo> =>
    request.post('/tables/scan', { scene }),

  getAvailableTables: (storeId: number, peopleCount = 2): Promise<TableItem[]> =>
    request.get('/tables/available', { params: { store_id: storeId, people_count: peopleCount } }),

  checkin: (data: { tableId: number; peopleCount: number; reservationId?: number; memberId?: number }) =>
    request.post('/tables/checkin', data),

  checkout: (data: { tableId: number; orderId: number }) =>
    request.post('/tables/checkout', data),
}

export const reservationApi = {
  create: (data: {
    storeId: number
    tableId?: number
    memberId?: number
    memberName: string
    memberPhone: string
    tableNo?: string
    reserveDate: string
    reserveTime: string
    peopleCount: number
    remark?: string
    source?: string
  }): Promise<ReservationItem> => request.post('/reservations', data),

  cancel: (id: number) =>
    request.post(`/reservations/${id}/cancel`),

  get: (id: number): Promise<ReservationItem> =>
    request.get(`/reservations/${id}`),

  list: (params: {
    storeId?: number
    memberId?: number
    status?: number
    reserveDate?: string
    checkinStatus?: number
    keyword?: string
    pageNum?: number
    pageSize?: number
  }) => request.get('/reservations', { params }),

  getTimeSlots: (params: {
    storeId: number
    reserveDate: string
    peopleCount?: number
  }): Promise<TimeSlot[]> => request.get('/reservations/timeslots', { params }),
}

export const queueApi = {
  create: (data: {
    storeId: number
    queueType?: string
    memberId?: number
    memberName: string
    memberPhone: string
    peopleCount: number
    remark?: string
  }): Promise<QueueItem> => request.post('/queues', data),

  cancel: (data: { queueId: number; reason?: string }) =>
    request.post('/queues/cancel', data),

  getStatus: (params: {
    storeId: number
    queueType?: string
    memberId?: number
    queueNumber?: string
  }): Promise<QueueStatus> => request.get('/queues/status', { params }),

  getMy: (params: { memberId: number; storeId: number }): Promise<QueueItem[]> =>
    request.get('/queues/my', { params }),

  getWaitingCount: (storeId: number): Promise<{ small: number; medium: number; large: number }> =>
    request.get('/queues/waiting-count', { params: { store_id: storeId } }),

  getConfig: (storeId: number): Promise<QueueConfig> =>
    request.get('/queue-config', { params: { store_id: storeId } }),
}
